src/System.Text.Primitives/System/Text/Encoding/Encoders.Utf16.cs

```diff
@@ -0,0 +1,790 @@
+                }
+            }
+
+            /// <summary>
+            /// Converts a span containing a sequence of UTF-8 bytes into UTF-16 bytes.
+            ///
+            /// This method will consume as many of the input bytes as possible.
+            ///
+            /// On successful exit, the entire input was consumed and encoded successfully. In this case, <paramref name="bytesConsumed"/> will be
+            /// equal to the length of the <paramref name="source"/> and <paramref name="bytesWritten"/> will equal the total number of bytes written to
+            /// the <paramref name="destination"/>.
+            /// </summary>
+            /// <param name="source">A span containing a sequence of UTF-8 bytes.</param>
+            /// <param name="destination">A span to write the UTF-16 bytes into.</param>
+            /// <param name="bytesConsumed">On exit, contains the number of bytes that were consumed from the <paramref name="source"/>.</param>
+            /// <param name="bytesWritten">On exit, contains the number of bytes written to <paramref name="destination"/></param>
+            /// <returns>A <see cref="TransformationStatus"/> value representing the state of the conversion.</returns>
+            public unsafe static TransformationStatus ConvertFromUtf8(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten)
+            {
+                fixed (byte* pUtf8 = &source.DangerousGetPinnableReference())
+                fixed (byte* pUtf16 = &destination.DangerousGetPinnableReference())
+                {
+                    byte* pSrc = pUtf8;
+                    byte* pSrcEnd = pSrc + source.Length;
+                    char* pDst = (char*)pUtf16;
+                    char* pDstEnd = pDst + (destination.Length >> 1);   // Conversion from bytes to chars - div by sizeof(char)
+
+                    int ch = 0;
+                    while (pSrc < pSrcEnd && pDst < pDstEnd)
+                    {
+                        // we may need as many as 1 character per byte, so reduce the byte count if necessary.
+                        // If availableChars is too small, pStop will be before pTarget and we won't do fast loop.
+                        int availableChars = EncodingHelper.PtrDiff(pDstEnd, pDst);
+                        int availableBytes = EncodingHelper.PtrDiff(pSrcEnd, pSrc);
+
+                        if (availableChars < availableBytes)
+                            availableBytes = availableChars;
+
+                        // don't fall into the fast decoding loop if we don't have enough bytes
+                        if (availableBytes <= 13)
+                        {
+                            // try to get over the remainder of the ascii characters fast though
+                            byte* pLocalEnd = pSrc + availableBytes;
+                            while (pSrc < pLocalEnd)
+                            {
+                                ch = *pSrc;
```

Comment:
Please file an issue to explore using SIMD widening operations to implement the ASCII to UTF16 part. Thanks!